srtp/srtp.c

```diff
@@ -2074,6 +2060,154 @@ srtp_remove_stream(srtp_t session, uint32_t ssrc) {
 }
 
 
+srtp_err_status_t
+srtp_update(srtp_t session, const srtp_policy_t *policy) {
+  srtp_err_status_t stat;
+
+  /* sanity check arguments */
+  if ((session == NULL) || (policy == NULL) || (policy->key == NULL))
+    return srtp_err_status_bad_param;
+
+  while (policy != NULL) {
+    stat = srtp_update_stream(session, policy);
+    if (stat) {
+      return stat;
+    }
+
+    /* set policy to next item in list  */
+    policy = policy->next;
+  }
+  return srtp_err_status_ok;
+}
+
+
+srtp_err_status_t
+srtp_update_stream(srtp_t session, const srtp_policy_t *policy) {
+  srtp_err_status_t status;
+
+  /* sanity check arguments */
+  if ((session == NULL) || (policy == NULL) || (policy->key == NULL))
+    return srtp_err_status_bad_param;
+
+  switch (policy->ssrc.type) {
+  case (ssrc_any_outbound):
+  case (ssrc_any_inbound): {
+    if (session->stream_template == NULL) {
+      return srtp_err_status_bad_param;
+    }
+
+    srtp_stream_t new_stream_template;
+    srtp_stream_t new_stream_list = NULL;
+
+    /* allocate new template stream  */
+    status = srtp_stream_alloc(&new_stream_template, policy);
+    if (status) {
+      return status;
+    }
+
+    /* initialize new template stream  */
+    status = srtp_stream_init(new_stream_template, policy);
+    if (status) {
+      srtp_crypto_free(new_stream_template);
+      return status;
+    }
+
+    /* for all old templated streams */
+    for (;;) {
+      srtp_stream_t stream = session->stream_list;
+      while ((stream != NULL) && (stream->rtp_auth != session->stream_template->rtp_auth)) {
+        stream = stream->next;
+      }
+      if (stream == NULL) {
+        /* no more templated streams */
+        break;
```

Comment:
breaking here is not an error, new_stream_template  will be used after this for loop so not sure why you think it would leak. line 2170.

All though if srtp_stream_dealloc fails on line 2136 and line 2151 fail then it will be leaked, but at that point there could be many leaks so not sure. I saw there was comment some where that if deallocating fails then memory will leaked.